modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/prepare/LeastRestrictiveTypesTest.java

```diff
@@ -0,0 +1,257 @@
+        tests.add(Arguments.arguments(INTEGER, SMALLINT, new LeastRestrictiveType(INTEGER)));
+        tests.add(Arguments.arguments(INTEGER, INTEGER, new LeastRestrictiveType(INTEGER)));
+        tests.add(Arguments.arguments(INTEGER, FLOAT, new LeastRestrictiveType(FLOAT)));
+        tests.add(Arguments.arguments(INTEGER, REAL, new LeastRestrictiveType(REAL)));
+        tests.add(Arguments.arguments(INTEGER, DOUBLE, new LeastRestrictiveType(DOUBLE)));
+        tests.add(Arguments.arguments(INTEGER, DECIMAL, new LeastRestrictiveType(DECIMAL)));
+        tests.add(Arguments.arguments(INTEGER, BIGINT, new LeastRestrictiveType(BIGINT)));
+
+        return tests.stream();
+    }
+
+    @ParameterizedTest
+    @MethodSource("floatTests")
+    public void testFloat(RelDataType t1, RelDataType t2, LeastRestrictiveType leastRestrictiveType) {
+        if (t1.equals(FLOAT) && t2.equals(DOUBLE)) {
+            // Something is not right here.
+            expectLeastRestrictiveType(t1, t2, new LeastRestrictiveType(FLOAT));
+            expectLeastRestrictiveType(t2, t1, new LeastRestrictiveType(DOUBLE));
+        } else {
+            expectLeastRestrictiveType(t1, t2, leastRestrictiveType);
+            expectLeastRestrictiveType(t2, t1, leastRestrictiveType);
+        }
+    }
+
+    private static Stream<Arguments> floatTests() {
+        List<Arguments> tests = new ArrayList<>();
+
+        tests.add(Arguments.arguments(FLOAT, TINYINT, new LeastRestrictiveType(FLOAT)));
+        tests.add(Arguments.arguments(FLOAT, SMALLINT, new LeastRestrictiveType(FLOAT)));
+        tests.add(Arguments.arguments(FLOAT, INTEGER, new LeastRestrictiveType(FLOAT)));
+        tests.add(Arguments.arguments(FLOAT, FLOAT, new LeastRestrictiveType(FLOAT)));
+        tests.add(Arguments.arguments(FLOAT, REAL, new LeastRestrictiveType(FLOAT)));
+        // See testFloat
+        tests.add(Arguments.arguments(FLOAT, DOUBLE, new LeastRestrictiveType(DOUBLE)));
+        tests.add(Arguments.arguments(FLOAT, DECIMAL, new LeastRestrictiveType(DOUBLE)));
+        tests.add(Arguments.arguments(FLOAT, BIGINT, new LeastRestrictiveType(FLOAT)));
+
+        return tests.stream();
+    }
+
+    @ParameterizedTest
+    @MethodSource("doubleTests")
+    public void testDouble(RelDataType t1, RelDataType t2, LeastRestrictiveType leastRestrictiveType) {
+        expectLeastRestrictiveType(t1, t2, leastRestrictiveType);
+        expectLeastRestrictiveType(t2, t1, leastRestrictiveType);
+    }
+
+    private static Stream<Arguments> doubleTests() {
+        List<Arguments> tests = new ArrayList<>();
+
+        tests.add(Arguments.arguments(DOUBLE, TINYINT, new LeastRestrictiveType(DOUBLE)));
+        tests.add(Arguments.arguments(DOUBLE, SMALLINT, new LeastRestrictiveType(DOUBLE)));
+        tests.add(Arguments.arguments(DOUBLE, INTEGER, new LeastRestrictiveType(DOUBLE)));
+        //See testFloat
+        //tests.add(Arguments.arguments(DOUBLE, FLOAT, new LeastRestrictiveType(FLOAT)));
```

Comment:
since this line is commented out, does it make sense to put the type we really expecting (`DOUBLE`)?

Reply:
Fixed